src/wasmtime/wasmtime.cc

```diff
@@ -0,0 +1,696 @@
+  uint32_t n = 0;
+  uint32_t shift = 0;
+  byte_t b;
+  do {
+    if (pos + 1 > end) {
+      abort();
+    }
+    b = *pos++;
+    n += (b & 0x7f) << shift;
+    shift += 7;
+  } while ((b & 0x80) != 0);
+  return n;
+}
+
+bool Wasmtime::load(const std::string &code, bool allow_precompiled) {
+  source_ = code;
+  store_ = wasm_store_new(engine());
+  // Wasm file header is 8 bytes (magic number + version).
+  static const uint8_t magic_number[4] = {0x00, 0x61, 0x73, 0x6d};
+  if (source_.size() < 8 || ::memcmp(source_.data(), magic_number, 4) != 0) {
+    return false;
+  }
+
+  WasmByteVec source_vec;
+  getStrippedSource(&source_vec);
+  module_ = wasm_module_new(store_.get(), source_vec.get());
+
+  if (module_) {
+    shared_module_ = wasm_module_share(module_.get());
+    assert(shared_module_ != nullptr);
+  }
+
+  return module_ != nullptr;
+}
+
+std::unique_ptr<WasmVm> Wasmtime::clone() {
+  assert(shared_module_ != nullptr);
+  auto clone = std::make_unique<Wasmtime>();
+
+  clone->integration().reset(integration()->clone());
+  clone->store_ = wasm_store_new(engine());
+  clone->module_ = wasm_module_obtain(clone->store_.get(), shared_module_.get());
+  return clone;
+}
+
+// TODO(mathetake): move to proxy_wasm::common::*
+bool Wasmtime::getStrippedSource(WasmByteVec *out) {
+  std::vector<byte_t> stripped;
+
+  const byte_t *pos = source_.data() + 8 /* Wasm header */;
+  const byte_t *end = source_.data() + source_.size();
+  while (pos < end) {
+    const auto section_start = pos;
+    if (pos + 1 > end) {
+      return false;
+    }
+    const auto section_type = *pos++;
+    const auto section_len = parseVarint(pos, end);
+    if (section_len == static_cast<uint32_t>(-1) || pos + section_len > end) {
+      return false;
+    }
+    if (section_type == 0 /* custom section */) {
+      const auto section_data_start = pos;
+      const auto section_name_len = parseVarint(pos, end);
+      if (section_name_len == static_cast<uint32_t>(-1) || pos + section_name_len > end) {
+        return false;
+      }
+      auto section_name = std::string_view(pos, section_name_len);
+      if (section_name.find("precompiled_") != std::string::npos) {
+        // If this is the first "precompiled_" section, then save everything
+        // before it, otherwise skip it.
+        if (stripped.empty()) {
+          const byte_t *start = source_.data();
+          stripped.insert(stripped.end(), start, section_start);
+        }
+      }
+      pos = section_data_start + section_len;
+    } else {
+      pos += section_len;
+      // Save this section if we already saw a custom "precompiled_" section.
+      if (!stripped.empty()) {
+        stripped.insert(stripped.end(), section_start, pos /* section end */);
+      }
+    }
+  }
+
+  if (stripped.empty()) {
+    wasm_byte_vec_new_uninitialized(out->get(), source_.size());
+    ::memcpy(out->get()->data, source_.data(), source_.size());
```

Comment:
This is unnecessary copy, which then persists in memory. If there are no changes, you should just return the unmodified bytecode.

Reply:
so is this what you want https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/73/commits/f7eecd3c5e095817963f0a0596f7b3dfa6db7390 ?

Reply:
No, this also makes a copy, just in a different place. Take a look at code in V8.

Basically, you need something that either creates new vector (if we stripped any sections) or returns original source (not a copy of the original source) or returns false and you use original source in the caller, if that's easier.